lang/py/src/avro/io.py

```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
 
 def validate(expected_schema, datum):
   """Determine if a python datum is an instance of a schema."""
+  BASE_ERR = 'Expected "{}" found {}'
   schema_type = expected_schema.type
   if schema_type == 'null':
-    return datum is None
+    return ValidationResult(
+      None if datum is None else BASE_ERR.format('null', type(datum))
+    )
   elif schema_type == 'boolean':
-    return isinstance(datum, bool)
+    return ValidationResult(
+      None if isinstance(datum, bool) else BASE_ERR.format('boolean', type(datum))
+    )
   elif schema_type == 'string':
-    return isinstance(datum, basestring)
+    return ValidationResult(
+      None if isinstance(datum, basestring) else BASE_ERR.format('string', type(datum))
+    )
   elif schema_type == 'bytes':
     if (hasattr(expected_schema, 'logical_type') and
-        expected_schema.logical_type == constants.DECIMAL):
-      return isinstance(datum, Decimal)
-    return isinstance(datum, str)
+      expected_schema.logical_type == constants.DECIMAL):
+      return ValidationResult(
+        None if isinstance(datum, Decimal) else \
+        BASE_ERR.format('logicalType(Decimal), baseType(bytes)', type(datum))
+      )
+    return ValidationResult(None if isinstance(datum, str) else BASE_ERR.format('bytes', type(datum)))
   elif schema_type == 'int':
     if hasattr(expected_schema, 'logical_type'):
       if expected_schema.logical_type == constants.DATE:
-        return isinstance(datum, datetime.date)
+        return ValidationResult(
+          None if isinstance(datum, datetime.date) else \
+          BASE_ERR.format('logicalType(Date), baseType(datetime.date)', type(datum))
+        )
       elif expected_schema.logical_type == constants.TIME_MILLIS:
-        return isinstance(datum, datetime.time)
-    return (isinstance(datum, (int, long))
-            and INT_MIN_VALUE <= datum <= INT_MAX_VALUE)
+        return ValidationResult(
+          None if isinstance(datum, datetime.time) else \
+          BASE_ERR.format('logicalType(Time-Millis), baseType(datetime.time)', type(datum))
+        )
+    return ValidationResult(
+      None if (isinstance(datum, (int, long)) and INT_MIN_VALUE <= datum <= INT_MAX_VALUE) else \
```

Comment:
nitpick: it may be easier to read if using a local variable to hold the result of checking, i.e.
```
is_valid = isinstance(...) and INT_MIN_VALUE <= datum <= INT_MAX_VALUE
return ValidationResult() if is_valid else ...
```